collect_app/src/main/java/org/odk/collect/android/activities/FormEntryActivity.java

```diff
@@ -1723,7 +1705,7 @@ private boolean saveForm(boolean exit, boolean complete, String updatedSaveName,
                              boolean current) {
         // save current answer
         if (current) {
-            if (!saveAnswersForCurrentScreen(complete)) {
+            if (!evaluateConstraintsAndSaveAnswersForCurrentScreen()) {
```

Comment:
Is this ok that in this case we always evaluate constraints?

Reply:
This looks like a mistake to me now. We're probably missing test coverage for this. Will need to investigate a bit more.

Reply:
Pulling in @lognaturel here as well as this is some nitty-gritty form behaviour stuff. I wouldn't be surprised if some of this is "bugs that are now features" behaviour.

After digging into this a little more, it definitely looks like we have pretty bad branch coverage for `saveForm`. On master, I can take the original implementation:

```java
// save current answer
if (current) {
    if (!saveAnswersForCurrentScreen(complete)) {
        showShortToast(this, R.string.data_saved_error);
        return false;
    }
}

formSaveViewModel.saveForm(getIntent().getData(), complete, updatedSaveName, exit);

return true;
```

...and change it to:

```java
saveAnswersForCurrentScreen(false)
formSaveViewModel.saveForm(getIntent().getData(), complete, updatedSaveName, exit);
return true;
```

I can pass `true` or `false` to `saveAnswersForCurrentScreen` - we only currently verify that the latest answers are saved when hitting the quick save icon or saving on quit (in `QuickSaveTest` and `QuitFormTest` respectively). Focusing on the different behaviours we should get from different values of `complete` (which seems to mean "finalized" really), I think the important behaviours around constraints we need to verify in tests are:

* A form can be quick saved/saved on quit with invalid constraints when it is not finalized (filling for the first time or editing a non finalized form)
* A form cannot be quick saved/saved on quit with invalid constraints when it has been finalized already (editing a finalized form)

There's a [third use of `saveForm`](https://github.com/getodk/collect/blob/5eb93ca5499a50d1d704ab3552377fb0df4279b3/collect_app/src/main/java/org/odk/collect/android/activities/FormEntryActivity.java#L1577) I was very confused about. Looking at the [Xforms spec](https://getodk.github.io/xforms-spec/#bind-attributes) it seems like this is basically a "quick save" that happens when the user views a certain question?  However, that use always passes `complete` as `false` which means it doesn't neatly fit into my stated behaviours above (as a "quick save"). Does it make sense that `saveIncomplete` allows saving without validating constraints on a finalized instance? If we want to maintain that behaviour then we need to also verify these:

* A form can be auto saved when a user views a `saveIncomplete` with invalid constraints when it is not finalized (filling for the first time or editing a non finalized form)
* A form can be auto saved when a user views a `saveIncomplete` with invalid constraints when it is finalized (editing a finalized form)

Reply:
I have no idea if anyone is using that `saveIncomplete` attribute and I'm not sure if it's needed at all but we should rather keep it especially we shouldn't get rid of it in this pr I think.

Reply:
`saveIncomplete` analytics are [here](https://console.firebase.google.com/u/2/project/api-project-322300403941/analytics/app/android:org.odk.collect.android/events/~2Freport~2FWidgetAttribute%3Ft%3D1657145869419&fpn%3D322300403941&swu%3D1&sgu%3D1&sus%3Dupgraded&params%3D_r..eventId%253DWidgetAttribute&cs%3Dapp.m.events.detail&r%3Ddefault&g%3D1). https://github.com/getodk/collect/issues/3758 has some information about why the feature exists.

> Does it make sense that saveIncomplete allows saving without validating constraints on a finalized instance?

My guess is that that part of the behavior is not critical. My understanding is that it used to be important to make sure data was committed as a submission (not as a savepoint file) at a certain point because an external app would read it. Now, it's still convenient to guarantee that data is saved as an instance (which is easier to deal with than a savepoint) but it's much less critical. Even in that old workflow I'm guessing edits were not very frequent. That said, if we were to change it, would it just silently fail if constraints failed? I don't think any other behavior would make sense. The other quick save cases involve explicit user action so they can send the user to fix failed constraints. I'm fine with either introducing the two tests you suggest or finding a coherent way to make the behavior better fit the "quick save" concept.

Reply:
Thanks for the thoughts folks. Seems like we're all on the same page of not being fans of `saveIncomplete`, but resigning ourselves to the fact that we probably need to keep the feature intact. I'll add tests for these behaviours:

1. A form can be quick saved/saved on quit with invalid constraints when it is not finalized (filling for the first time or editing a non finalized form)
1. A form cannot be quick saved/saved on quit with invalid constraints when it has been finalized already (editing a finalized form)
1. A form can be auto saved when a user views a saveIncomplete with invalid constraints when it is not finalized (filling for the first time or editing a non finalized form)
1. A form can be auto saved when a user views a saveIncomplete with invalid constraints when it is finalized (editing a finalized form)

I'll do this on master to check everything passes, then add them into this PR and getting them passing here should fix up the code this comment was original about.

Reply:
...and of course there's some extra complexity that we've missed in this discussion. There's different behaviour based on whether the violated constraint is on the "current" screen when we save as we evaluate constraints and update answers for the current screen first, then save the whole form to disk. I need to investigate further how constraints behave in the "save to disk" part.

Reply:
Ok the crux of it is that there is different "save with violated constraints" for finalized forms based on if the violated constraint is on the current screen or not. If the constraint is on the current screen, the user sees a toast that says something cryptic like "save failed" whereas if the constraint is on another screen, they are moved to the relevant screen, and they see a proper constraint error message ("question required" for instance). In the former case, the process that updates the current answers catches the constraint violation and in the latter it's the save to disk that gets it. This doesn't seem ideal at all, but it is the current behaviour so I'm writing up tests to capture it.